service/matching/matching_engine_test.go

```diff
@@ -2761,6 +2762,321 @@ func (s *matchingEngineSuite) TestUpdatePhysicalTaskQueueGauge_BuildID() {
+	// We shall manually remove the task from the testTaskManager and then call AckManager.CompleteTask(TaskID) to see the resetting take place
+
+	// TODO Shivam - This should fail with the current implementation. Have a skip or something.
+
+	s.matchingEngine.config.LongPollExpirationInterval = dynamicconfig.GetDurationPropertyFnFilteredByTaskQueueInfo(10 * time.Millisecond)
+	_, workflowExecution := s.generateWorkflowExecution()
+
+	const taskCount = 1
+	const taskID = 11
+
+	taskQueue, ptq := s.createTQAndPTQForBacklogTests()
+	s.addWorkflowTasks(false, 1, taskCount, taskQueue, workflowExecution, nil)
+	s.EqualValues(taskCount, s.taskManager.getTaskCount(ptq))
+
+	// validating the approximateBacklogCounter
+	tlMgr := s.getPhysicalTaskQueueManagerImpl(ptq)
+	s.EqualValues(tlMgr.backlogMgr.db.getApproximateBacklogCount(), int64(1))
+
+	// Remove the task from testTaskManager but not from db/AckManager
+	request := &persistence.CompleteTasksLessThanRequest{
+		NamespaceID:        namespaceId,
+		TaskQueueName:      taskQueue.Name,
+		TaskType:           enumspb.TASK_QUEUE_TYPE_WORKFLOW,
+		ExclusiveMaxTaskID: taskID + 1,
+		Limit:              100,
+	}
+	_, err := s.taskManager.CompleteTasksLessThan(context.Background(), request)
+	s.NoError(err)
+	// Decreasing the backlogCounter since a task has been removed in theory
+	tlMgr.backlogMgr.db.updateApproximateBacklogCount(int64(-1))
+
+	// overriding DB's logger to catch ERROR's
+	mockLogger := log.NewMockLogger(s.controller)
+	mockLogger.EXPECT().Error("ApproximateBacklogCounter could have under-counted. This should never happen",
+		tag.WorkerBuildId(tlMgr.backlogMgr.db.queue.BuildId()), tag.WorkerBuildId(tlMgr.backlogMgr.db.queue.Partition().NamespaceId().String()))
+	tlMgr.backlogMgr.db.logger = mockLogger
+	_, ok := tlMgr.backlogMgr.db.logger.(*log.MockLogger)
+	s.Equal(ok, true)
+
+	// Call AckManager.CompleteTasks with the same TaskID
+	tlMgr.backlogMgr.taskAckManager.completeTask(int64(taskID))
+
+	// Log outputs an error since we would have under-counted. Counter is not reset.
+	s.EqualValues(tlMgr.backlogMgr.db.getApproximateBacklogCount(), int64(0))
+
+}
+
+func (s *matchingEngineSuite) TestUnloadingTQMValidateBacklogCounter() {
+
+	s.matchingEngine.config.LongPollExpirationInterval = dynamicconfig.GetDurationPropertyFnFilteredByTaskQueueInfo(10 * time.Millisecond)
+	_, workflowExecution := s.generateWorkflowExecution()
+	taskQueue, ptq := s.createTQAndPTQForBacklogTests()
+	const taskCount = 20
+
+	s.addWorkflowTasks(false, 1, taskCount, taskQueue, workflowExecution, nil)
+	s.EqualValues(taskCount, s.taskManager.getTaskCount(ptq))
+
+	tlMgr := s.getPhysicalTaskQueueManagerImpl(ptq)
+	s.EqualValues(tlMgr.backlogMgr.db.getApproximateBacklogCount(), taskCount)
+
+	tm, ok := tlMgr.backlogMgr.db.store.(*testTaskManager)
+	s.EqualValues(ok, true)
+	s.EqualValues(tm.getUpdateCount(ptq), math.Ceil(float64(taskCount)/float64(s.matchingEngine.config.RangeSize)))
```

Comment:
I think this check was taken from a previously written test case but I don't seem to think it's really required here, given that I do verify the backlog counter above. 

Having said that, the `Update` call happens because:

1. We allocate ID's based on the `RangeSize` (10, in this case). 
2. We are adding 20 tasks, which results in calling the function to allocate a block twice.
3. When (2) happens, the function `renewTaskQueueLocked` which in turn calls `UpdateState`.